tests/consensus_tests/test_resharding.py

```diff
@@ -0,0 +1,332 @@
+        # Update existing points on the first peer
+        run_in_background(upsert_points_throttled, peer_api_uris[0], COLLECTION_NAME, start=0, end=num_upserts),
+        # Delete points on the first two peers
+        run_in_background(delete_points_throttled, peer_api_uris[0], COLLECTION_NAME, start=500, end=500 + num_deletes),
+        run_in_background(delete_points_throttled, peer_api_uris[1], COLLECTION_NAME, start=600, end=600 + num_deletes),
+    ]
+
+    # Reshard 3 times in sequence
+    for shard_count in range(2, 5):
+        # Start resharding
+        r = requests.post(
+            f"{peer_api_uris[0]}/collections/{COLLECTION_NAME}/cluster", json={
+                "start_resharding": {
+                    "peer_id": first_peer_id,
+                    "shard_key": None,
+                }
+            })
+        assert_http_ok(r)
+
+        # Wait for resharding operation to start and stop
+        wait_for_collection_resharding_operations_count(peer_api_uris[0], COLLECTION_NAME, 1)
+        for uri in peer_api_uris:
+            wait_for_collection_resharding_operations_count(uri, COLLECTION_NAME, 0)
+
+        # Assert node shard count
+        for uri in peer_api_uris:
+            assert check_collection_local_shards_count(uri, COLLECTION_NAME, shard_count)
+
+    # Wait for background updates to finish
+    for task in update_tasks:
+        while task.is_alive():
+            pass
+
+    # Assert node shard and point sum count
+    # Expects base points + 3x upserts - 2x deletes
+    expected_points = num_points + num_upserts * 3 - num_deletes * 2
+    for uri in peer_api_uris:
+        assert check_collection_local_shards_point_count(uri, COLLECTION_NAME, expected_points)
+
+    sleep(1)
+
+    # Match all points on all nodes exactly
+    # Note: due to concurrent updates on all peers this check may fail, but I've
+    # not seen this yet. Once it does, we probably want to remove this.
+    data = []
+    for uri in peer_api_uris:
+        r = requests.post(
+            f"{uri}/collections/{COLLECTION_NAME}/points/scroll", json={
+                "limit": 999999999,
+                "with_vectors": True,
+                "with_payload": True,
+            }
+        )
+        assert_http_ok(r)
+        data.append(r.json()["result"])
+    check_data_consistency(data)
+
+
+def run_in_background(run, *args, **kwargs):
+    p = multiprocessing.Process(target=run, args=args, kwargs=kwargs)
+    p.start()
+    return p
+
+
+def upsert_points_throttled(peer_url, collection_name, start=0, end=None):
+    batch_size = 3
+    offset = start
+
+    while True:
+        count = min(end - offset, batch_size) if end is not None else batch_size
+        if count <= 0:
+            return
+
+        upsert_random_points(peer_url, count, collection_name, offset=offset)
+        offset += count
+
+        sleep(random.uniform(0.4, 0.6))
```

Comment:
One small insertion once in half a second look a bit low. How long is the whole transfer usually takes?

Reply:
This is sourced from WAL delta tests, where more uploads would be come flaky in some cases because of very slow CI.

I'll speed it up a bit for now. We can lower upsertions again if it becomes problematic.

Reply:
We agreed on doing this in a separate PR not to stall this one any longer.